get_smb.go

```diff
@@ -0,0 +1,320 @@
+	// check if file exists in the smb shared folder and check the mode
+	isDir, err := isDirectory(baseCmd, file)
+	if err != nil {
+		return 0, err
+	}
+	if isDir {
+		return ModeDir, nil
+	}
+	return ModeFile, nil
+}
+
+func (g *SmbGetter) Get(ctx context.Context, req *Request) error {
+	if req.u.Host == "" || req.u.Path == "" {
+		return fmt.Errorf(basePathError)
+	}
+
+	// If dst folder doesn't exists, we need to remove the created on later in case of failures
+	dstExisted := false
+	if req.Dst != "" {
+		if _, err := os.Lstat(req.Dst); err == nil {
+			dstExisted = true
+		}
+	}
+
+	// First look in a possible local mount of the shared folder
+	path := "/" + req.u.Host + req.u.Path
+	if runtime.GOOS == "windows" {
+		path = "/" + path
+	}
+	f := new(FileGetter)
+	result := f.get(path, req)
+	if result == nil {
+		return nil
+	}
+
+	// If not mounted, try downloading the directory content using smbclient cli
+	err := g.smbclientGet(req)
+	if err == nil {
+		return nil
+	}
+
+	result = multierror.Append(result, err)
+
+	if !dstExisted {
+		// Remove the destination created for smbclient
+		os.Remove(req.Dst)
+	}
+
+	return fmt.Errorf("one of the options should be available: \n 1. local mount of the smb shared folder or; \n 2. smbclient cli installed (provice credentials when necessary). \n err: %s", result.Error())
+}
+
+func (g *SmbGetter) smbclientGet(req *Request) error {
+	hostPath, directory, err := findHostAndFilePath(req.u)
+	if err != nil {
+		return err
+	}
+
+	baseCmd := smbclientBaseCmd(req.u.User, hostPath, ".")
+	// check directory exists in the smb shared folder and is a directory
+	isDir, err := isDirectory(baseCmd, directory)
+	if err != nil {
+		return err
+	}
+	if !isDir {
+		return fmt.Errorf("%s source path must be a directory", directory)
+	}
+
+	// download everything that's inside the directory (files and subdirectories)
+	smbclientCmd := baseCmd + " --command 'prompt OFF;recurse ON; mget *'"
+
+	if req.Dst != "" {
+		_, err := os.Lstat(req.Dst)
+		if err != nil {
+			if os.IsNotExist(err) {
+				// Create destination folder if it doesn't exist
+				if err := os.MkdirAll(req.Dst, 0755); err != nil {
+					return fmt.Errorf("failed to create destination path: %s", err.Error())
+				}
+			} else {
+				return err
+			}
+		}
+	}
+
+	_, err = runSmbClientCommand(smbclientCmd, req.Dst)
+	return err
+}
+
+func (g *SmbGetter) GetFile(ctx context.Context, req *Request) error {
+	if req.u.Host == "" || req.u.Path == "" {
+		return fmt.Errorf(basePathError)
+	}
+
+	// If dst folder doesn't exist, we need to remove the created one later in case of failures
+	dstExisted := false
+	if req.Dst != "" {
+		if _, err := os.Lstat(req.Dst); err == nil {
+			dstExisted = true
+		}
+	}
+
+	// First look in a possible local mount of the shared folder
+	path := "/" + req.u.Host + req.u.Path
+	if runtime.GOOS == "windows" {
+		path = "/" + path
+	}
+	f := new(FileGetter)
+	result := f.getFile(path, req, ctx)
+	if result == nil {
```

Comment:
@SwampDragons This is where the url is transformed from `smb://`  to `//` or `/`, depending on the OS, and used by the FileGetter.